### DISCUSSION TITLE
Firebase Messaging handling APNS token change

### DISCUSSION THREAD

**Discussion by aryaxt** [firebase/firebase-ios-sdk#10685] [Q&A]
APNS token can change time to time. If i have firebase messaging installed with swizzling enabled, does it call registerForRemoteNotifications time to time in case the token has changed? or is that something I need to do myself?

Should I add logic myself to call registerForRemoteNotifications if authorizationStatus is authorized? or is that done by Messaging?

**Comment by heoseongjin** [comment]
> APNS token can change time to time. If i have firebase messaging installed with swizzling enabled, does it call registerForRemoteNotifications time to time in case the token has changed? or is that something I need to do myself?
> 
> Should I add logic myself to call registerForRemoteNotifications if authorizationStatus is authorized? or is that done by Messaging?

- Firebase Messaging automatically handles APNs token changes when swizzling is enabled, so you generally don't need to manually call `registerForRemoteNotifications`. 
- When the user's authorization status is authorized, Messaging takes care of it. Implement the `messaging:didReceiveRegistrationToken:` method to handle token updates.

😆

**Reply by shasranj** [reply]
1. I tried to call this but fcm token is not getting generated:
FirebaseApp.configure()
Messaging.messaging().delegate = self
        
UNUserNotificationCenter.current().requestAuthorization(options: [.sound, .alert, .badge]) { (isPermissionGranted, error) in
}

2. But these below method is not getting called.
APNS:
func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
    // Not getting called
}
FCM:
func messaging(_ messaging: Messaging, didReceiveRegistrationToken fcmToken: String?) {
    // Not getting called
}

So are you sure we don't need to call  application.registerForRemoteNotifications() in didFinishLaunchingWithOptions ???
When i am adding these line only then token is getting generated.
